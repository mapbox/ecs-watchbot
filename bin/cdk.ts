#!/usr/bin/env node
import 'source-map-support/register';
import {Construct} from "constructs";
import {
    App,
    aws_codebuild,
    aws_codepipeline,
    aws_codestarconnections,
    BootstraplessSynthesizer,
    CliCredentialsStackSynthesizer,
    Duration,
    RemovalPolicy,
    Stack,
    StackProps,
} from "aws-cdk-lib";
import {Bucket, BucketEncryption} from "aws-cdk-lib/aws-s3";
import {BuildEnvironmentVariableType, BuildSpec} from "aws-cdk-lib/aws-codebuild";
import {LogGroup, RetentionDays} from "aws-cdk-lib/aws-logs";
import {Effect, Policy, PolicyDocument, PolicyStatement, Role, ServicePrincipal} from "aws-cdk-lib/aws-iam";

const app = new App();

// const input =  CodePipelineSource.connection('mapbox/ecs-watchbot', 'master', {
//     connectionArn: 'arn:aws:codestar-connections:us-east-1:353802256504:connection/da887df2-781b-4a00-8dce-24dced5cc129',
// });

interface Props extends StackProps {
    deploymentEnvironment: string
    bucketName: string
}

class PipelineStack extends Stack {
    constructor(scope: Construct, id: string, props: Props) {
        super(scope, id, props);

        const codebuildLogGroup = new LogGroup(this, 'CodebuildLogs', {
            logGroupName: `${this.stackName}-logs`,
            retention: RetentionDays.TWO_WEEKS,
            removalPolicy: RemovalPolicy.DESTROY,
        });

        const linuxCodebuild = new aws_codebuild.PipelineProject(this, 'LinuxBundle', {
            environment: {
                buildImage: aws_codebuild.LinuxBuildImage.STANDARD_7_0
            },
            logging: {
                cloudWatch: {
                    enabled: true,
                    logGroup: codebuildLogGroup,
                }
            },
            environmentVariables: {
                BUCKET_NAME: {
                    type: BuildEnvironmentVariableType.PLAINTEXT,
                    value: props.bucketName
                }
            },
            projectName: `${this.stackName}-linux-bundle`,
            buildSpec: BuildSpec.fromObject({
                version: '0.2',
                phases: {
                    install: {
                        'commands': [
                            'npm ci --production'
                        ]
                    },
                    build: {
                        commands:[
                            'node bin/watchbot-binary-generator',
                        ],
                    },
                },
            })
        });
        const alpineCodebuild = new aws_codebuild.PipelineProject(this, 'AlpineBundle', {
            environment: {
                buildImage: aws_codebuild.LinuxBuildImage.fromCodeBuildImageId('public.ecr.aws/docker/library/node:18-alpine')
            },
            logging: {
                cloudWatch: {
                    enabled: true,
                    logGroup: codebuildLogGroup,
                }
            },
            environmentVariables: {
                BUCKET_NAME: {
                    type: BuildEnvironmentVariableType.PLAINTEXT,
                    value: props.bucketName
                }
            },
            projectName: `${this.stackName}-alpine-bundler`,
            description: 'Builds watchbot binaries for alpine OS',
            buildSpec: BuildSpec.fromObject({
                version: '0.2',
                phases: {
                    install: {
                        'commands': [
                            'apk add git',
                            'npm ci --production'
                        ]
                    },
                    build: {
                        commands:[
                            'node bin/watchbot-binary-generator alpine',
                        ],
                    },
                },
            })
        })

        const s3Permissions = new Policy(this, 'S3Policy', {
            statements: [new PolicyStatement({
                effect: Effect.ALLOW,
                actions: [
                    "s3:PutObject",
                    "s3:GetObject",
                    "s3:GetObjectVersion"
                ],
                resources: [
                    `arn:aws:s3:::${props.bucketName}`,
                    `arn:aws:s3:::${props.bucketName}/*`,
                ]
            })]
        })
        linuxCodebuild.role?.attachInlinePolicy(s3Permissions);
        alpineCodebuild.role?.attachInlinePolicy(s3Permissions);

        const role = new Role(this, 'PipelineRole', {
            assumedBy: new ServicePrincipal('codepipeline.amazonaws.com'),
            inlinePolicies: {
                main: new PolicyDocument({
                    statements: [new PolicyStatement({
                        sid: 'S3permissions',
                        effect: Effect.ALLOW,
                        actions: [
                            's3:*'
                        ],
                        resources: [
                            `arn:aws:s3:::${props.bucketName}`,
                            `arn:aws:s3:::${props.bucketName}/*`,
                        ]
                    }), new PolicyStatement({
                        effect: Effect.ALLOW,
                        actions: [
                            'codebuild:StartBuild',
                            'codebuild:BatchGetBuilds',
                            'codestar-connections:*'
                        ],
                        resources: ['*']
                    }), new PolicyStatement({
                        effect: Effect.ALLOW,
                        actions: [
                            'sts:AssumeRole'
                        ],
                        resources: [
                            linuxCodebuild.role?.roleArn || '',
                            alpineCodebuild.role?.roleArn  || '',
                        ]
                    })]
                })
            }
        });

        /*
         aws_codepipeline.Pipeline creates v1 which can't be re-triggered with overrides and that's needed for testing
         codepipeline.CodePipeline creates v2 but forces you to have a stack
        */
        const connection = new aws_codestarconnections.CfnConnection(this, 'Connection', {
            connectionName: 'ecs-watchbot',
            providerType: 'GitHub'
        });
        const p = new aws_codepipeline.CfnPipeline(this, 'Pipeline', {
            roleArn: role.roleArn,
            name: this.stackName,
            stages: [{
                name: 'Source',
                actions: [{
                    name: 'Github',
                    actionTypeId: {
                        category: 'Source',
                        owner: 'AWS',
                        provider: 'CodeStarSourceConnection',
                        version: '1',
                    },
                    configuration: {
                        ConnectionArn: connection.attrConnectionArn,
                        FullRepositoryId: 'mapbox/ecs-watchbot',
                        BranchName: 'master',
                    },
                    outputArtifacts: [{
                        name: 'Source',
                    }],
                }],
            }, {
                name: 'Bundle',
                actions: [{
                    name: 'Bundle',
                    actionTypeId: {
                        category: 'Build',
                        owner: 'AWS',
                        provider: 'CodeBuild',
                        version: '1',
                    },
                    inputArtifacts: [{
                        name: 'Source'
                    }],
                    configuration: {
                        ProjectName: linuxCodebuild.projectName
                    }
                }, {
                    name: 'AlpineBundle',
                    actionTypeId: {
                        category: 'Build',
                        owner: 'AWS',
                        provider: 'CodeBuild',
                        version: '1',
                    },
                    inputArtifacts: [{
                        name: 'Source'
                    }],
                    configuration: {
                        ProjectName: alpineCodebuild.projectName
                    }
                }]
            }],
            artifactStore: {
                location: props.bucketName,
                type: 'S3',
            },
        });
        p.addPropertyOverride('PipelineType', 'V2');
        p.addPropertyOverride('Triggers', [
            {
                ProviderType: 'CodeStarSourceConnection',
                GitConfiguration: {
                    SourceActionName: 'Github',
                    Push: [{
                        Tags: {
                            Includes: ['*.*.*-*']
                        }
                    }]
                }
            }
        ]);
    }
}


class BucketStack extends Stack {
    constructor(scope: Construct, id: string, props: Props) {
        super(scope, id, props);

        new Bucket(this, 'Bucket', {
            bucketName: props.bucketName,
            blockPublicAccess: {
                blockPublicAcls: true,
                ignorePublicAcls: true,
                blockPublicPolicy: false,
                restrictPublicBuckets: false,
            },
            versioned: true,
            encryption: BucketEncryption.S3_MANAGED,
            lifecycleRules: [{
                enabled: true,
                abortIncompleteMultipartUploadAfter: Duration.days(1),
            }]
        });

    }

}

const isProduction = (env: string) => env === 'production';
const tags = {
    Team: 'DataPlatform',
    ServiceOrganization: 'Platform',
    Classification: 'internal',
    Public: 'false',
}
const region: string = process.env.AWS_DEFAULT_REGION || 'us-east-1';
const deploymentEnvironment = app.node.tryGetContext('deploymentEnvironment');
const account = app.node.tryGetContext('account');
const bucketName = isProduction(deploymentEnvironment) ? 'ecs-watchbot-binaries' : 'ecs-watchbot-binaries-stg';

const pipelineStackName = 'watchbot-pipeline';
new PipelineStack(app, 'Pipeline', {
    stackName: pipelineStackName,
    synthesizer: new CliCredentialsStackSynthesizer({ // TODO give cdk roles codestar-connections:* actions
       fileAssetsBucketName: `cdk-assets-${account}-${region}`,
        bucketPrefix: `ecs-watchbot/${pipelineStackName}`,
       qualifier: 'operator'
    }),
    // synthesizer: new BootstraplessSynthesizer({
    //     cloudFormationExecutionRoleArn: process.env.AWS_CDK_EXEC_ROLE,
    //     deployRoleArn: process.env.AWS_CDK_DEPLOY_ROLE,
    // }),
    env: { account, region },
    tags: {
        ...tags,
        CloudFormationStackName: pipelineStackName,
        Production: isProduction(deploymentEnvironment).toString(),
    },
    deploymentEnvironment: deploymentEnvironment,
    bucketName,
});

const bucketStackName = 'watchbot-bucket';
new BucketStack(app, 'Bucket', {
    stackName: bucketStackName,
    synthesizer: new BootstraplessSynthesizer({
        cloudFormationExecutionRoleArn: process.env.AWS_CDK_EXEC_ROLE,
        deployRoleArn: process.env.AWS_CDK_DEPLOY_ROLE,
    }),
    env: { account, region },
    tags: {
        ...tags,
        CloudFormationStackName: bucketStackName,
        Production: isProduction(deploymentEnvironment).toString(),
    },
    deploymentEnvironment: deploymentEnvironment,
    bucketName,
});
