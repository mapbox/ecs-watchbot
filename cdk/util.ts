import {App} from "aws-cdk-lib";

export const tryGetContext = (app: App, context: string) => {
    const value = app.node.tryGetContext(context);
    if (!value) throw new Error(`Context -c ${context} is missing.`);
    return value;
}

export const isProduction = (env: string) => env === 'production';
