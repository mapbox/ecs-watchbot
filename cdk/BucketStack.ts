import {Duration, Stack, StackProps} from "aws-cdk-lib";
import {Construct} from "constructs";
import {BlockPublicAccess, Bucket, BucketEncryption} from "aws-cdk-lib/aws-s3";
import {isProduction} from "./util";

interface Props extends StackProps {
    deploymentEnvironment: string
    bucketName: string
}

export class BucketStack extends Stack {
    constructor(scope: Construct, id: string, props: Props) {
        super(scope, id, props);

        const bucket = new Bucket(this, 'Bucket', {
            bucketName: props.bucketName,
            publicReadAccess: true,
            blockPublicAccess: BlockPublicAccess.BLOCK_ACLS,
            versioned: true,
            encryption: BucketEncryption.S3_MANAGED,
            lifecycleRules: [{
                enabled: true,
                id: 'AbortIncompleteMultipartUploads',
                abortIncompleteMultipartUploadAfter: Duration.days(1),
            }, {
                enabled: true,
                id: 'ExpireNonCurrentVersions',
                noncurrentVersionExpiration: Duration.days(1),
            }]
        });

        if (!isProduction(props.deploymentEnvironment)) { // delete test versions only
            bucket.addLifecycleRule({
                expiration: Duration.days(14)
            });
        }

    }

}
