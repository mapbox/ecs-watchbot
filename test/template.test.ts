import { App, Stack } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { PipelineStack } from '../cdk/PipelineStack';
import { BucketStack } from '../cdk/BucketStack';

describe('Template', () => {
  let stack: Stack;
  let template: Template;

  const dummyStackName = 'dummy-stack';
  const stackProps = {
    stackName: dummyStackName,
    env: { account: '222258372212', region: 'us-east-1' },
    tags: {
      Team: 'DataPlatform',
      ServiceOrganization: 'Platform',
      Classification: 'internal',
      Public: 'false',
      CloudFormationStackName: dummyStackName,
      Production: 'false'
    },
    deploymentEnvironment: 'staging',
    bucketName: 'ecs-watchbot-binaries-stg'
  };

  describe('When creating the pipeline stack', () => {
    beforeEach(() => {
      stack = new PipelineStack(new App(), 'Pipeline', stackProps);
      template = Template.fromStack(stack);
    });

    it('Creates Pipeline to match snapshot', () => {
      expect(template).toMatchSnapshot();
    });
  });

  describe('When creating the bucket stack', () => {
    beforeEach(() => {
      stack = new BucketStack(new App(), 'Bucket', stackProps);
      template = Template.fromStack(stack);
    });

    it('Creates Bucket to match snapshot', () => {
      expect(template).toMatchSnapshot();
    });
  });
});
