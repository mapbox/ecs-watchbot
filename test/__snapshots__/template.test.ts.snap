// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Template When creating the bucket stack Creates Bucket to match snapshot 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "Bucket83908E77": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "BucketEncryption": {
          "ServerSideEncryptionConfiguration": [
            {
              "ServerSideEncryptionByDefault": {
                "SSEAlgorithm": "AES256",
              },
            },
          ],
        },
        "BucketName": "ecs-watchbot-binaries-stg",
        "LifecycleConfiguration": {
          "Rules": [
            {
              "AbortIncompleteMultipartUpload": {
                "DaysAfterInitiation": 1,
              },
              "Id": "AbortIncompleteMultipartUploads",
              "Status": "Enabled",
            },
            {
              "Id": "ExpireNonCurrentVersions",
              "NoncurrentVersionExpiration": {
                "NoncurrentDays": 1,
              },
              "Status": "Enabled",
            },
            {
              "ExpirationInDays": 14,
              "Status": "Enabled",
            },
          ],
        },
        "PublicAccessBlockConfiguration": {
          "BlockPublicAcls": true,
          "IgnorePublicAcls": true,
        },
        "VersioningConfiguration": {
          "Status": "Enabled",
        },
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "BucketPolicyE9A3008A": {
      "Properties": {
        "Bucket": {
          "Ref": "Bucket83908E77",
        },
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "s3:GetObject",
              "Effect": "Allow",
              "Principal": {
                "AWS": "*",
              },
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    {
                      "Fn::GetAtt": [
                        "Bucket83908E77",
                        "Arn",
                      ],
                    },
                    "/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::S3::BucketPolicy",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;

exports[`Template When creating the pipeline stack Creates Pipeline to match snapshot 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "AlpineBundle646137AC": {
      "Properties": {
        "Artifacts": {
          "Type": "CODEPIPELINE",
        },
        "Cache": {
          "Type": "NO_CACHE",
        },
        "Description": "Builds watchbot binaries for alpine OS",
        "EncryptionKey": "alias/aws/s3",
        "Environment": {
          "ComputeType": "BUILD_GENERAL1_SMALL",
          "EnvironmentVariables": [
            {
              "Name": "BUCKET_NAME",
              "Type": "PLAINTEXT",
              "Value": "ecs-watchbot-binaries-stg",
            },
          ],
          "Image": "public.ecr.aws/docker/library/node:18-alpine",
          "ImagePullCredentialsType": "CODEBUILD",
          "PrivilegedMode": false,
          "Type": "LINUX_CONTAINER",
        },
        "LogsConfig": {
          "CloudWatchLogs": {
            "GroupName": {
              "Ref": "CodebuildLogs71A917B2",
            },
            "Status": "ENABLED",
          },
        },
        "Name": "dummy-stack-alpine-bundler",
        "ServiceRole": {
          "Fn::GetAtt": [
            "AlpineBundleRole73C0A50F",
            "Arn",
          ],
        },
        "Source": {
          "BuildSpec": "{
  "version": "0.2",
  "phases": {
    "install": {
      "commands": [
        "apk add git",
        "npm ci --production"
      ]
    },
    "build": {
      "commands": [
        "node bin/watchbot-binary-generator alpine"
      ]
    }
  }
}",
          "Type": "CODEPIPELINE",
        },
      },
      "Type": "AWS::CodeBuild::Project",
    },
    "AlpineBundleRole73C0A50F": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "codebuild.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "AlpineBundleRoleDefaultPolicy4B149A9D": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "logs:CreateLogStream",
                "logs:PutLogEvents",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "CodebuildLogs71A917B2",
                  "Arn",
                ],
              },
            },
            {
              "Action": [
                "logs:CreateLogGroup",
                "logs:CreateLogStream",
                "logs:PutLogEvents",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":logs:us-east-1:222258372212:log-group:/aws/codebuild/",
                      {
                        "Ref": "AlpineBundle646137AC",
                      },
                    ],
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":logs:us-east-1:222258372212:log-group:/aws/codebuild/",
                      {
                        "Ref": "AlpineBundle646137AC",
                      },
                      ":*",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": [
                "codebuild:CreateReportGroup",
                "codebuild:CreateReport",
                "codebuild:UpdateReport",
                "codebuild:BatchPutTestCases",
                "codebuild:BatchPutCodeCoverages",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:",
                    {
                      "Ref": "AWS::Partition",
                    },
                    ":codebuild:us-east-1:222258372212:report-group/",
                    {
                      "Ref": "AlpineBundle646137AC",
                    },
                    "-*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "AlpineBundleRoleDefaultPolicy4B149A9D",
        "Roles": [
          {
            "Ref": "AlpineBundleRole73C0A50F",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "CodebuildLogs71A917B2": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "LogGroupName": "dummy-stack-logs",
        "RetentionInDays": 14,
      },
      "Type": "AWS::Logs::LogGroup",
      "UpdateReplacePolicy": "Delete",
    },
    "Connection": {
      "Properties": {
        "ConnectionName": "ecs-watchbot",
        "ProviderType": "GitHub",
      },
      "Type": "AWS::CodeStarConnections::Connection",
    },
    "LinuxBundleFEBE4AAA": {
      "Properties": {
        "Artifacts": {
          "Type": "CODEPIPELINE",
        },
        "Cache": {
          "Type": "NO_CACHE",
        },
        "EncryptionKey": "alias/aws/s3",
        "Environment": {
          "ComputeType": "BUILD_GENERAL1_SMALL",
          "EnvironmentVariables": [
            {
              "Name": "BUCKET_NAME",
              "Type": "PLAINTEXT",
              "Value": "ecs-watchbot-binaries-stg",
            },
          ],
          "Image": "aws/codebuild/standard:7.0",
          "ImagePullCredentialsType": "CODEBUILD",
          "PrivilegedMode": false,
          "Type": "LINUX_CONTAINER",
        },
        "LogsConfig": {
          "CloudWatchLogs": {
            "GroupName": {
              "Ref": "CodebuildLogs71A917B2",
            },
            "Status": "ENABLED",
          },
        },
        "Name": "dummy-stack-linux-bundle",
        "ServiceRole": {
          "Fn::GetAtt": [
            "LinuxBundleRole857A200D",
            "Arn",
          ],
        },
        "Source": {
          "BuildSpec": "{
  "version": "0.2",
  "phases": {
    "install": {
      "commands": [
        "npm ci --production"
      ]
    },
    "build": {
      "commands": [
        "node bin/watchbot-binary-generator"
      ]
    }
  }
}",
          "Type": "CODEPIPELINE",
        },
      },
      "Type": "AWS::CodeBuild::Project",
    },
    "LinuxBundleRole857A200D": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "codebuild.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "LinuxBundleRoleDefaultPolicyF6C6CB6D": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "logs:CreateLogStream",
                "logs:PutLogEvents",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "CodebuildLogs71A917B2",
                  "Arn",
                ],
              },
            },
            {
              "Action": [
                "logs:CreateLogGroup",
                "logs:CreateLogStream",
                "logs:PutLogEvents",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":logs:us-east-1:222258372212:log-group:/aws/codebuild/",
                      {
                        "Ref": "LinuxBundleFEBE4AAA",
                      },
                    ],
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":logs:us-east-1:222258372212:log-group:/aws/codebuild/",
                      {
                        "Ref": "LinuxBundleFEBE4AAA",
                      },
                      ":*",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": [
                "codebuild:CreateReportGroup",
                "codebuild:CreateReport",
                "codebuild:UpdateReport",
                "codebuild:BatchPutTestCases",
                "codebuild:BatchPutCodeCoverages",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:",
                    {
                      "Ref": "AWS::Partition",
                    },
                    ":codebuild:us-east-1:222258372212:report-group/",
                    {
                      "Ref": "LinuxBundleFEBE4AAA",
                    },
                    "-*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "LinuxBundleRoleDefaultPolicyF6C6CB6D",
        "Roles": [
          {
            "Ref": "LinuxBundleRole857A200D",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "Pipeline": {
      "Properties": {
        "ArtifactStore": {
          "Location": "ecs-watchbot-binaries-stg",
          "Type": "S3",
        },
        "Name": "dummy-stack",
        "PipelineType": "V2",
        "RoleArn": {
          "Fn::GetAtt": [
            "PipelineRoleDCFDBB91",
            "Arn",
          ],
        },
        "Stages": [
          {
            "Actions": [
              {
                "ActionTypeId": {
                  "Category": "Source",
                  "Owner": "AWS",
                  "Provider": "CodeStarSourceConnection",
                  "Version": "1",
                },
                "Configuration": {
                  "BranchName": "master",
                  "ConnectionArn": {
                    "Fn::GetAtt": [
                      "Connection",
                      "ConnectionArn",
                    ],
                  },
                  "FullRepositoryId": "mapbox/ecs-watchbot",
                },
                "Name": "Github",
                "OutputArtifacts": [
                  {
                    "Name": "Source",
                  },
                ],
              },
            ],
            "Name": "Source",
          },
          {
            "Actions": [
              {
                "ActionTypeId": {
                  "Category": "Build",
                  "Owner": "AWS",
                  "Provider": "CodeBuild",
                  "Version": "1",
                },
                "Configuration": {
                  "ProjectName": {
                    "Ref": "TestCodebuild44B59119",
                  },
                },
                "InputArtifacts": [
                  {
                    "Name": "Source",
                  },
                ],
                "Name": "Test",
              },
            ],
            "Name": "Test",
          },
          {
            "Actions": [
              {
                "ActionTypeId": {
                  "Category": "Build",
                  "Owner": "AWS",
                  "Provider": "CodeBuild",
                  "Version": "1",
                },
                "Configuration": {
                  "ProjectName": {
                    "Ref": "LinuxBundleFEBE4AAA",
                  },
                },
                "InputArtifacts": [
                  {
                    "Name": "Source",
                  },
                ],
                "Name": "Bundle",
              },
              {
                "ActionTypeId": {
                  "Category": "Build",
                  "Owner": "AWS",
                  "Provider": "CodeBuild",
                  "Version": "1",
                },
                "Configuration": {
                  "ProjectName": {
                    "Ref": "AlpineBundle646137AC",
                  },
                },
                "InputArtifacts": [
                  {
                    "Name": "Source",
                  },
                ],
                "Name": "AlpineBundle",
              },
            ],
            "Name": "Bundle",
          },
        ],
      },
      "Type": "AWS::CodePipeline::Pipeline",
    },
    "PipelineRoleDCFDBB91": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "codepipeline.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Policies": [
          {
            "PolicyDocument": {
              "Statement": [
                {
                  "Action": "s3:*",
                  "Effect": "Allow",
                  "Resource": [
                    "arn:aws:s3:::ecs-watchbot-binaries-stg",
                    "arn:aws:s3:::ecs-watchbot-binaries-stg/*",
                  ],
                  "Sid": "S3permissions",
                },
                {
                  "Action": [
                    "codebuild:StartBuild",
                    "codebuild:BatchGetBuilds",
                    "codestar-connections:*",
                  ],
                  "Effect": "Allow",
                  "Resource": "*",
                },
                {
                  "Action": "sts:AssumeRole",
                  "Effect": "Allow",
                  "Resource": [
                    {
                      "Fn::GetAtt": [
                        "LinuxBundleRole857A200D",
                        "Arn",
                      ],
                    },
                    {
                      "Fn::GetAtt": [
                        "AlpineBundleRole73C0A50F",
                        "Arn",
                      ],
                    },
                  ],
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "main",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "S3Policy8FACFAB8": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "s3:PutObject",
                "s3:GetObject",
                "s3:GetObjectVersion",
              ],
              "Effect": "Allow",
              "Resource": [
                "arn:aws:s3:::ecs-watchbot-binaries-stg",
                "arn:aws:s3:::ecs-watchbot-binaries-stg/*",
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "S3Policy8FACFAB8",
        "Roles": [
          {
            "Ref": "LinuxBundleRole857A200D",
          },
          {
            "Ref": "AlpineBundleRole73C0A50F",
          },
          {
            "Ref": "TestCodebuildRoleD0274C56",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "TestCodebuild44B59119": {
      "Properties": {
        "Artifacts": {
          "Type": "CODEPIPELINE",
        },
        "Cache": {
          "Type": "NO_CACHE",
        },
        "Description": "Runs tests",
        "EncryptionKey": "alias/aws/s3",
        "Environment": {
          "ComputeType": "BUILD_GENERAL1_SMALL",
          "EnvironmentVariables": [
            {
              "Name": "BUCKET_NAME",
              "Type": "PLAINTEXT",
              "Value": "watchbot-binaries",
            },
          ],
          "Image": "aws/codebuild/standard:7.0",
          "ImagePullCredentialsType": "CODEBUILD",
          "PrivilegedMode": false,
          "Type": "LINUX_CONTAINER",
        },
        "LogsConfig": {
          "CloudWatchLogs": {
            "GroupName": {
              "Ref": "CodebuildLogs71A917B2",
            },
            "Status": "ENABLED",
          },
        },
        "Name": "dummy-stack-test",
        "ServiceRole": {
          "Fn::GetAtt": [
            "TestCodebuildRoleD0274C56",
            "Arn",
          ],
        },
        "Source": {
          "BuildSpec": "{
  "version": "0.2",
  "phases": {
    "pre_build": {
      "commands": [
        "docker build -q -t ecs-watchbot -f test/Dockerfile ./"
      ]
    },
    "build": {
      "commands": [
        "docker run -t ecs-watchbot npm run test-container"
      ]
    }
  }
}",
          "Type": "CODEPIPELINE",
        },
      },
      "Type": "AWS::CodeBuild::Project",
    },
    "TestCodebuildRoleD0274C56": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "codebuild.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "TestCodebuildRoleDefaultPolicy9B3918A7": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "logs:CreateLogStream",
                "logs:PutLogEvents",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "CodebuildLogs71A917B2",
                  "Arn",
                ],
              },
            },
            {
              "Action": [
                "logs:CreateLogGroup",
                "logs:CreateLogStream",
                "logs:PutLogEvents",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":logs:us-east-1:222258372212:log-group:/aws/codebuild/",
                      {
                        "Ref": "TestCodebuild44B59119",
                      },
                    ],
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":logs:us-east-1:222258372212:log-group:/aws/codebuild/",
                      {
                        "Ref": "TestCodebuild44B59119",
                      },
                      ":*",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": [
                "codebuild:CreateReportGroup",
                "codebuild:CreateReport",
                "codebuild:UpdateReport",
                "codebuild:BatchPutTestCases",
                "codebuild:BatchPutCodeCoverages",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:",
                    {
                      "Ref": "AWS::Partition",
                    },
                    ":codebuild:us-east-1:222258372212:report-group/",
                    {
                      "Ref": "TestCodebuild44B59119",
                    },
                    "-*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "TestCodebuildRoleDefaultPolicy9B3918A7",
        "Roles": [
          {
            "Ref": "TestCodebuildRoleD0274C56",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
